File: sdkwork-im-sdk-typescript/composed/src/media-module.ts
import type {
  AttachMediaRequest,
  AttachTextMediaOptions,
  ImMediaUploadOptions,
  ImMediaUploadSession,
  ImUploadedMediaAsset,
  CompleteUploadRequest,
  CreateUploadRequest,
  MediaAsset,
  MediaDownloadUrlResponse,
  PostMessageResult,
  QueryParams,
} from './types.js';
import type { ImSdkContext } from './sdk-context.js';
import {
  normalizeUploadSession,
  performPresignedMediaUpload,
} from './media-upload-runtime.js';

export class ImMediaModule {
  constructor(private readonly context: ImSdkContext) {}

  async createUpload(body: CreateUploadRequest): Promise<ImMediaUploadSession> {
    const response = await this.context.transportClient.media.createMediaUpload(body);
    return normalizeUploadSession(response);
  }

  createUploadSession(body: CreateUploadRequest): Promise<ImMediaUploadSession> {
    return this.createUpload(body);
  }

  completeUpload(
    mediaAssetId: string | number,
    body: CompleteUploadRequest,
  ): Promise<MediaAsset> {
    return this.context.transportClient.media.completeMediaUpload(mediaAssetId, body);
  }

  uploadAndComplete(
    options: ImMediaUploadOptions,
  ): Promise<ImUploadedMediaAsset> {
    return this.upload(options);
  }

  async upload(options: ImMediaUploadOptions): Promise<ImUploadedMediaAsset> {
    return performPresignedMediaUpload(this.context, options);
  }

  getDownloadUrl(
    mediaAssetId: string | number,
    params?: QueryParams,
  ): Promise<MediaDownloadUrlResponse> {
    return this.context.transportClient.media.getMediaDownloadUrl(mediaAssetId, params);
  }

  get(mediaAssetId: string | number): Promise<MediaAsset> {
    return this.context.transportClient.media.getMediaAsset(mediaAssetId);
  }

  attach(
    mediaAssetId: string | number,
    body: AttachMediaRequest,
  ): Promise<PostMessageResult> {
    return this.context.transportClient.media.attachMediaAsset(mediaAssetId, body);
  }

  attachText(
    mediaAssetId: string | number,
    options: AttachTextMediaOptions,
  ): Promise<PostMessageResult> {
    return this.attach(mediaAssetId, {
      conversationId: options.conversationId,
      clientMsgId: options.clientMsgId,
      summary: options.summary,
      text: options.text,
      renderHints: options.renderHints,
    });
  }
}

File: sdkwork-im-sdk-rust/composed/src/media_module.rs
use crate::{ImSdkContext, QueryParams};
use sdkwork_im_sdk_generated::{
  AttachMediaRequest,
  CompleteUploadRequest,
  CreateUploadRequest,
  MediaAsset,
  MediaDownloadUrlResponse,
  PostMessageResult,
  SdkworkError,
};

#[derive(Clone)]
pub struct ImMediaModule {
  context: ImSdkContext,
}

impl ImMediaModule {
  pub(crate) fn new(context: ImSdkContext) -> Self {
    Self { context }
  }

  pub async fn create_upload(
    &self,
    body: CreateUploadRequest,
  ) -> Result<MediaAsset, SdkworkError> {
    self.context.transport_client().media().create_media_upload(&body).await
  }

  pub async fn complete_upload(
    &self,
    media_asset_id: impl AsRef<str>,
    body: CompleteUploadRequest,
  ) -> Result<MediaAsset, SdkworkError> {
    let media_asset_id = media_asset_id.as_ref().to_string();
    self
      .context
      .transport_client()
      .media()
      .complete_media_upload(&media_asset_id, &body)
      .await
  }

  pub async fn download_url(
    &self,
    media_asset_id: impl AsRef<str>,
    params: Option<&QueryParams>,
  ) -> Result<MediaDownloadUrlResponse, SdkworkError> {
    let media_asset_id = media_asset_id.as_ref().to_string();
    self
      .context
      .transport_client()
      .media()
      .get_media_download_url(&media_asset_id, params)
      .await
  }

  pub async fn get(
    &self,
    media_asset_id: impl AsRef<str>,
  ) -> Result<MediaAsset, SdkworkError> {
    let media_asset_id = media_asset_id.as_ref().to_string();
    self.context.transport_client().media().get_media_asset(&media_asset_id).await
  }

  pub async fn attach(
    &self,
    media_asset_id: impl AsRef<str>,
    body: AttachMediaRequest,
  ) -> Result<PostMessageResult, SdkworkError> {
    let media_asset_id = media_asset_id.as_ref().to_string();
    self
      .context
      .transport_client()
      .media()
      .attach_media_asset(&media_asset_id, &body)
      .await
  }
}

